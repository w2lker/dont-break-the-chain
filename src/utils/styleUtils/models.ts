export type cssAttribute = 'unset' | 'null' | number | string | undefined | null;

type TArrayWithLength<TItem, TLength extends number> = [TItem, ...TItem[]] & { length: TLength };
export type BoundAttributeArray = TArrayWithLength<string, 1| 2 | 3 | 4>;
