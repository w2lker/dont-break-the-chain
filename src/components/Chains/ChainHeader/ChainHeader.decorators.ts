import withStyles from '@material-ui/core/styles/withStyles';

import ChainHeader from './ChainHeader';
import chainHeaderStyles from './ChainHeader.styles';

// @ts-ignore
const styledComponent = withStyles(chainHeaderStyles)(ChainHeader);

export default styledComponent;
