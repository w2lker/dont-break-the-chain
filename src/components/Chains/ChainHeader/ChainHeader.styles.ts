import { styleUtils } from '../../../utils/styleUtils';

const chainHeaderStyles = {
  wrapper: {
    ...styleUtils.minMaxWidth(170),
    ...styleUtils.setFlexContainer('flex-start', 'center'),
    paddingTop: 10,
    paddingBottom: 11,
  },
  icon: {
    fontSize: 20,
  },
  description: {
    marginLeft: 15,
  },
  name: {
    // TODO: define colors as constants
    ...styleUtils.setFont(14, '#263238', 500, '0.1px'),
    margin: '0 0 15px',
    maxWidth: 135,
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
  },
  stats: {
    ...styleUtils.setFont(12, '#7A8F9B', 500, '0.1px'),
  },
};

export default chainHeaderStyles;
