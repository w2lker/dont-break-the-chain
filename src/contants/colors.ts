import colorBlue from '@material-ui/core/colors/blue';
import colorBrown from '@material-ui/core/colors/brown';
import colorCyan from '@material-ui/core/colors/cyan';
import colorDeepPurple from '@material-ui/core/colors/deepPurple';
import colorGreen from '@material-ui/core/colors/green';
import colorIndigo from '@material-ui/core/colors/indigo';
import colorOrange from '@material-ui/core/colors/orange';
import colorPink from '@material-ui/core/colors/pink';
import colorPurple from '@material-ui/core/colors/purple';
import colorTeal from '@material-ui/core/colors/teal';

export type colorsKeys = 'pink' | 'purple' | 'deepPurple' | 'indigo' | 'blue' | 'cyan' | 'teal' | 'green' | 'orange' | 'brown';

export const colors: {[color: string]: any}  = {
  blue: colorBlue,
  brown: colorBrown,
  pink: colorPink,
  cyan: colorCyan,
  deepPurple: colorDeepPurple,
  indigo: colorIndigo,
  green: colorGreen,
  orange: colorOrange,
  purple: colorPurple,
  teal: colorTeal,
};
