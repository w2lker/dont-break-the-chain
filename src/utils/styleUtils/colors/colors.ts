import {bonWarning} from '../helpers/warning/warning';

export interface IRgbColor {
  red: number;
  green: number;
  blue: number;
}

const black: IRgbColor = {
  red: 0,
  green: 0,
  blue: 0,
};

function regularHexToRgb(hex: string): IRgbColor {
  const [red, green, blue] = [0, 2, 4]
    .map(offset => hex.substring(offset, offset + 2))
    .map(hexCh => parseInt(hexCh, 16));
  return { red, green, blue };
}

export function hexToRgbMapped(hex?: string): IRgbColor {
  const warnType = 'HEX to RGB converter';
  if (!hex) {
    return black;
  }
  if (hex.length === 1) {
    return regularHexToRgb(hex.repeat(6));
  }

  if (hex.length === 2) {
    return regularHexToRgb(hex.repeat(3));
  }

  if (hex.length === 3) {
    const [r, g, b] = hex.split('') as string[];
    return regularHexToRgb(`${r}${r}${g}${g}${b}${b}`);
  }

  if (hex.length === 4 || hex.length === 5) {
    const hex3 = hex.slice(0, 3);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex3} value`, warnType);
    return hexToRgbMapped(hex3);
  }

  if (hex.length > 6) {
    const hex6 = hex.slice(0, 6);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex6} value`, warnType);
    return regularHexToRgb(hex6);
  }

  return regularHexToRgb(hex);
}
