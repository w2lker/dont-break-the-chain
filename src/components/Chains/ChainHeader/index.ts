import ChainHeader from "./ChainHeader";
export default ChainHeader;