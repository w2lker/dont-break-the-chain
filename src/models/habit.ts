import {List} from 'immutable';
import {colorsKeys} from "../contants/colors";

export const habitStatus = {
  incomplete: 0,
  complete: 1,
  paused: 2,
  start: 3,
  end: 4,
  incompleteToday: 5
};

export interface IHabitDate {
  date?: any;
  status?: string;
  score?: number;
}

export interface IHabit {
  id?: number;
  name?: string;
  color?: colorsKeys;
  icon?: string;
  currentChain?: number;
  longestChain?: number;
  dates?: List<IHabitDate>;
}