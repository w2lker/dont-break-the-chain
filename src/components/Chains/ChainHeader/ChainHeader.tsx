import * as React from 'react';

import MaterialDesignIcon from '../../basic/Material-Icon';

import { colors } from '../../../contants/colors';
import { IHabit } from '../../../models/habit';

const ChainHeader: React.FC<IHabit> = (props) => {
  const { currentChain, longestChain } = props;
  // Handling incorrect attributes (empty string etc)
  const icon = props.icon || 'ninja';
  const color = colors[props.color || 'blue'][500];
  const name = props.name || 'The habit';

  const areStatsProvided = currentChain && longestChain;
  // @ts-ignore
  const longestChainUpdated = areStatsProvided && (longestChain > currentChain) ? longestChain : currentChain;

  const renderStats = areStatsProvided ? (
    <span className="stats">
      {`${currentChain}/${longestChainUpdated}`}
    </span>
  ) : null;

  return (
    <div className="habit-header-wrapper">
      <MaterialDesignIcon
        name={icon}
        style={{ color }}
      />
      <div className="description">
        <h5 className="name">
          {name}
        </h5>
        {renderStats}
      </div>
    </div>
  );
};

export default ChainHeader;
