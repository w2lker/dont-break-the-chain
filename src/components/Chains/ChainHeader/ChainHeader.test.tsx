import { mount, render, shallow } from 'enzyme';
import { List } from 'immutable';
import React from 'react';

import ChainHeader from './ChainHeader';

import { colors } from '../../../contants/colors';
import { IHabit } from '../../../models/habit';

describe('Chain Header Component', () => {
  const propsEmpty: IHabit = {
    id: undefined,
    name: '',
    // @ts-ignore
    color: '',
    icon: '',
    currentChain: undefined,
    longestChain: undefined,
    dates: List(),
  };

  const propsSample: IHabit = {
    id: 1,
    name: 'Sample name',
    color: 'cyan',
    icon: 'sample',
    currentChain: 1,
    longestChain: 15,
    dates: List(),
  };

  it('match snapshot', () => {
    const component = mount(<ChainHeader {...propsSample} />);
    expect(component.debug()).toMatchSnapshot();
  });

  describe('testing default props', () => {
    const component = shallow(<ChainHeader {...propsEmpty} />);

    it('renders default name', () => {
      expect(component.find('.name').text()).toBe('The habit');
    });

    it('renders default icon props', () => {
      const defaultStyle = {
        color: colors.blue[500],
      };
      const defaultName = 'ninja';
      const iconComponent = component.find('MaterialDesignIcon');
      const iconProps = iconComponent.props();
      expect(iconProps.name).toBe(defaultName);
      expect(iconProps.style).toEqual(defaultStyle);
    });

    it('renders without stats', () => {
      expect(component.find('.stats')).toEqual({});
    });
  });

  describe('testing custom props', () => {
    const component = shallow(<ChainHeader {...propsSample} />);

    it('renders custom name', () => {
      expect(component.find('.name').text()).toBe(propsSample.name);
    });

    it('renders custom icon props', () => {
      const sampleStyle = {
        // @ts-ignore
        color: colors[propsSample.color][500],
      };
      const sampleName = propsSample.icon;
      const iconComponent = component.find('MaterialDesignIcon');
      const iconProps = iconComponent.props();
      expect(iconProps.name).toBe(sampleName);
      expect(iconProps.style).toEqual(sampleStyle);
    });

    it('renders custom stats', () => {
      const sampleStats = `${propsSample.currentChain}/${propsSample.longestChain}`;
      expect(component.find('.stats').text()).toEqual(sampleStats);
    });
  });

});
