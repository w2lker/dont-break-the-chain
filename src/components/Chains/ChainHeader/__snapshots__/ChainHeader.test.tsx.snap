// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chain Header Component match snapshot 1`] = `
"<ChainHeader id={1} name=\\"Sample name\\" color=\\"cyan\\" icon=\\"sample\\" currentChain={1} longestChain={15} dates={{...}}>
  <div className=\\"habit-header-wrapper\\">
    <MaterialDesignIcon name=\\"sample\\" style={{...}}>
      <span className=\\"material-icon mdi mdi-sample\\" style={{...}} />
    </MaterialDesignIcon>
    <div className=\\"description\\">
      <h5 className=\\"name\\">
        Sample name
      </h5>
      <span className=\\"stats\\">
        1/15
      </span>
    </div>
  </div>
</ChainHeader>"
`;
